pkg/resources/stage_grant.go

```diff
@@ -92,18 +94,32 @@ func StageGrant() *TerraformGrantResource {
 
 // CreateStageGrant implements schema.CreateFunc.
 func CreateStageGrant(d *schema.ResourceData, meta interface{}) error {
-	var stageName string
-	if name, ok := d.GetOk("stage_name"); ok {
-		stageName = name.(string)
-	}
+	var (
+		schemaName string
+		stageName  string
+	)
+
 	dbName := d.Get("database_name").(string)
-	schemaName := d.Get("schema_name").(string)
+	onFuture := d.Get("on_future").(bool)
 	priv := d.Get("privilege").(string)
-	futureStages := d.Get("on_future").(bool)
 	grantOption := d.Get("with_grant_option").(bool)
 
+	if name, ok := d.GetOk("schema_name"); ok {
+		schemaName = name.(string)
+	}
+	if name, ok := d.GetOk("stage_name"); ok {
+		stageName = name.(string)
+	}
+
+	if (schemaName == "") && !onFuture {
+		return errors.New("schema_name must be set unless on_future is true")
+	}
+	if (stageName == "") && !onFuture {
+		return errors.New("stage_name must be set unless on_future is true")
+	}
+
 	var builder snowflake.GrantBuilder
-	if futureStages {
+	if onFuture {
 		builder = snowflake.FutureStageGrant(dbName, schemaName)
 	} else {
 		builder = snowflake.StageGrant(dbName, schemaName, stageName)
```

Comment:
When setting grant, should we be passing roles to generate a unique grant id? I think that would be a breaking change if we added tho